flang/lib/Lower/Bridge.cpp

```diff
@@ -499,6 +504,101 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     }
   }
 
+  mlir::Value createThreadprivateVarClone(
+      const Fortran::semantics::Symbol &sym) override final {
+    auto loc = genLocation(sym.name());
+    Fortran::lower::SymbolBox sb = lookupSymbol(sym);
+    fir::ExtendedValue exv = sb.toExtendedValue();
+
+    auto allocate = [&](mlir::Type ty) -> mlir::Value {
+      return builder->allocateLocal(
+          loc, ty, mangleName(sym), toStringRef(sym.GetUltimate().name()),
+          /*pinned=*/true, fir::factory::getExtents(*builder, loc, exv),
+          fir::getTypeParams(exv),
+          sym.GetUltimate().attrs().test(Fortran::semantics::Attr::TARGET));
+    };
+
+    mlir::Value allocaVal = nullptr;
+    if (const auto *commonDet =
+            sym.detailsIf<Fortran::semantics::CommonBlockDetails>()) {
+      mlir::Type symType = getSymbolAddress(sym).getType();
+      auto refTy = symType.dyn_cast<fir::ReferenceType>();
+      symType = refTy.getEleTy();
+      allocaVal = allocate(symType);
+
+      for (const auto &mem : commonDet->objects()) {
+        const auto *memDet =
+            mem->detailsIf<Fortran::semantics::ObjectEntityDetails>();
+        assert(memDet != nullptr && "No common member object entity found");
+        Fortran::lower::SymbolBox memCsb = lookupSymbol(*mem);
+        auto *cvtSymOp = memCsb.getAddr().getDefiningOp();
+        auto *coordOp = cvtSymOp->getOperand(0).getDefiningOp();
+        auto cvtBlkVal = builder->createConvert(
+            loc, coordOp->getOperand(0).getType(), allocaVal);
+        auto coordVal = builder->create<fir::CoordinateOp>(
+            loc, cvtSymOp->getOperand(0).getType(), cvtBlkVal,
+            coordOp->getOperand(1));
+        auto cvtSymVal =
+            builder->createConvert(loc, memCsb.getAddr().getType(), coordVal);
+
+        fir::ExtendedValue memCexv = memCsb.toExtendedValue();
+        auto memExval = memCexv.match(
+            [&](const fir::CharBoxValue &box) -> fir::ExtendedValue {
+              return fir::CharBoxValue(
+                  cvtSymVal, fir::factory::readCharLen(*builder, loc, box));
+            },
+            [&](const fir::ArrayBoxValue &box) -> fir::ExtendedValue {
+              return fir::ArrayBoxValue{
+                  cvtSymVal, fir::factory::getExtents(*builder, loc, box)};
+            },
+            [&](const fir::CharArrayBoxValue &box) -> fir::ExtendedValue {
+              return fir::CharArrayBoxValue{
+                  cvtSymVal, fir::factory::readCharLen(*builder, loc, box),
+                  fir::factory::getExtents(*builder, loc, box)};
+            },
+            [&](const fir::MutableBoxValue &box) -> fir::ExtendedValue {
+              return fir::MutableBoxValue(cvtSymVal, box.nonDeferredLenParams(),
+                                          {});
+            },
+            [&](const auto &) -> fir::ExtendedValue { return cvtSymVal; });
+        localSymbols.addSymbol(*mem, memExval, true);
```

Comment:
If we have not created new symbols for threadprivate in the OpenMP region, wouldn't this replace the mlir::Value/fir::ExtendedValue associated with the original variable/symbol?

Reply:
It won't replace since the associations within and outside OpenMP region are in different symbolMapStack. The new symbolMapStack is generated [here](https://github.com/flang-compiler/f18-llvm-project/blob/851484444f1d2049e327837874040f3ac981af35/flang/lib/Lower/Bridge.cpp#L1560) when entering the OpenMP region, and all the statements using the symbols in the OpenMP region use this new symbolMapStack, and this symbolMapStack is popped after exiting the OpenMP region (check [here](https://github.com/flang-compiler/f18-llvm-project/blob/851484444f1d2049e327837874040f3ac981af35/flang/lib/Lower/Bridge.cpp#L1614-L1617)). So, all the mlir::Value/fir::ExtendedValue associated with the original variable/symbol are not affect outside the OpenMP region.